### ISSUE TITLE
Git Panel not detecting repo

### ISSUE THREAD

**Issue by mpilosov** [jupyterlab/jupyterlab-git#383]
![Screen Shot 2019-07-16 at 8 24 00 AM](https://user-images.githubusercontent.com/16293134/61302481-1415ed80-a7a3-11e9-9624-bd692b4e6bb7.png)

I tried the `--system` flag as in #202 but it did not work. Neither did adding/omitting `--sys-prefix` (It worked before, but then I upgraded to JLab 1.0), through DockerSpawner via JupyterHub. I waited to upgrade until I saw commits here referencing the compatibility, but the setup instructions as listed didn't work for me. The icon in the file browser that allows you to clone does gray out when I enter a repo, so the extension does seem to have _some_ awareness of its PATH, but the Panel appears to not reflect this. 

If I refresh the page, I can sometimes get the Git Panel to reflect the correct directory but it still shows the "Go Find a Repo" button.

**Comment by bernhard-42** [comment]
It looks like there are two prerequisites (found by debugging GitLab.js):
1) There must be an upstream origin
2) There must be at least one local tag
None of these seem to be created when using git init and commit in jupyterlab-git

**Comment by mathematicalmichael** [comment]
Are you sure this doesn't have anything to do with how the containers are set up? I think this problem happens even when there are upstreams.

**Comment by bernhard-42** [comment]
**Step 1:** Open jupyterlab in a folder not controlled by git
**Step 2:** Initialise git repo via jupyterlab Git menu
    => repo is visible in git sidebar
**Step3:** Commit files in the folder via the jupyterlab git sidebar
    => the files are committed (checked on command line), however a javascript error occurs and the files stay in "Staged" area despite being committed in the jupyterlab git ui

Error:
```
TypeError: Cannot read property 'length' of undefined
    at GitPanel.<anonymous> (vendors~main.e32d5fbe3838f9ee7447.js:339300)
    at Generator.next (<anonymous>)
    at fulfilled (vendors~main.e32d5fbe3838f9ee7447.js:339247)
```

![image](https://user-images.githubusercontent.com/954408/61587455-05587d80-ab8b-11e9-8ba1-64a3c8d04b53.png)

Which actually means that `apiResult.data.branch` is empty.

And this is basically the root cause:
If I set a breakpoint in GitPanel.js to look at the `apiResult`, I get

```
branch:
code: 0
command: "git show-ref"
message: "Error [fatal: Kein Upstream-Branch für Branch 'master' konfiguriert.↵] occurred while executing [git rev-parse --abbrev-ref master@{upstream}] command to get upstream branch."
```
which btw. means "No upstream for branch master configured"

**Step 4:** Let's restart jupyterlab (just for the sake of the argument)
![image](https://user-images.githubusercontent.com/954408/61587554-9e3bc880-ab8c-11e9-9ccf-da5e25f28593.png)

**Step 5:** Let's configure an upstream. 
Now the error message of the API call has changed:

```
branch:
code: 0
command: "git show-ref"
message: "Error [fatal: Keine Namen gefunden, kann nichts beschreiben.↵] occurred while executing [git describe --tags da607a604d04008ee259f5f394e260a3455f92f0] command to get nearest tag associated with branch."
```
which means "no names found cannot describe anything"

**Step 6** Let's create an annotated (!) tag

```
git tag v0 -m "first version"
```

And suddenly all works in jupyterlab git view.


So, no idea what you mean with "how the containers are set up", however, the above is how to reproduce the behaviour and build a workaround to work with the git extension.

**Comment by mpilosov** [comment]
interesting! 
well, that helped me look into the issue more. 
I was having trouble accessing a repo that had tags and upstreams and everything, and it wasn't showing up. A much more "new" repo I cloned seemed to work perfectly. 
I have a bunch that shows up in this problematic repo when I run `git show-ref`
when I navigate in the browser back to the problematic repo, the panel still shows the info from the previously selected repo. I had two remotes. Tried to delete one but it didn't help. 

So here's what I tried... I went into a different directory and re-cloned the repo... and suddenly, it appeared! So, this must have something to do with the repo's settings. What else might I try to investigate the difference between a fresh clone and this existing one? 

here is the config:
```
[core]
    repositoryformatversion = 0 
    filemode = true
    bare = false
    logallrefupdates = true
[branch "master"]
[remote "origin"]
    url = https://github.com/mathematicalmichael/BET.git
    fetch = +refs/heads/*:refs/remotes/origin/*
[branch "improvements"]
    remote = origin
    merge = refs/heads/improvements
[branch "linting"]
    remote = origin
    merge = refs/heads/linting
```

**Comment by bernhard-42** [comment]
I had to create annotated tags, not just tags.
The API call against git endpoint expected that else it gave an error for me

No idea how that aligns with your findings ...

**Comment by jaipreet-s** [comment]
Part of the issue is that the API which determines if the current path is a git repo is very "heavy". There's a need of lighter API here which doesn't have a dependency of tag or branch.

There was some work around that done in #350 to change the logic which disables or enables the Git Clone button if the current path is a Git repo.

**Comment by mayadebellis** [comment]
I am having this issue as well. It was fine before the upgrade to jupyterlab 1.0 but ever since then I am getting errors coming from GitPanel and it is not able to detect my git repos.

**Comment by jaipreet-s** [comment]
> I am having this issue as well. It was fine before the upgrade to jupyterlab 1.0 but ever since then I am getting errors coming from GitPanel and it is not able to detect my git repos.

Hi @mayadebellis : What do the logs from the Jupyter server say?

**Comment by oldnaari** [comment]
I am not sure if I shall report the issue here. I am also getting "Go find a repo". I assume it is because I use fish shell as default. When I open terminal using Git -> Open Terminal it starts with a message:
```
> cd ""
cd: Unknown error trying to locate directory “”
```

**Comment by oldnaari** [comment]
It seems it wasn't the reason. I have changed my default shell to bash and it still doesn't work.

**Comment by foxes1998** [comment]
Are you run jupyterlab as root ?

**Comment by foxes1998** [comment]
If you run jupyterlab as root, you should add a new general user, give him the Permission to your root directory and subdirectories ,then copy the `/root/.jupyter/` to his home directory, may you should copy the `/root/.bashrc` file too

** this solution may not the best one **

**Comment by oldnaari** [comment]
I run 3 instances of jupyter lab on two machines. Neither of them is run with sudo. Here are some insights on the situation.

| git plugin works? | machine | additional options | git repository      |
|------------------|---------|--------------------|---------------------|
| **yes**              | 1st     |                    | some private repo 1 |
| **no**               | 1st     |                    | some private repo 2 |
| **yes**              | 2nd    | --allow-root       | some private repo 2 |

**Comment by stefanvangastel** [comment]
Using latest (1.1.1 jupyterlab and 8.1 jupy-git) I can't get the sidepanel to detect my git repo in any way (tried tag solution etc.).

**Comment by VolokzhaninVadim** [comment]
> Using latest (1.1.1 jupyterlab and 8.1 jupy-git) I can't get the sidepanel to detect my git repo in any way (tried tag solution etc.).

I have the same conditions and problem.
![Снимок экрана от 2019-09-03 23-20-58](https://user-images.githubusercontent.com/27136123/64176860-a3857600-cea1-11e9-9e37-f7db46b1fca7.png)

**Comment by stefanvangastel** [comment]
In my case it was due to a misconfigured proxy, check your dev tools for any JS errors

**Comment by firasm** [comment]
My git repo wasn't being detected either on Jupyterlab 1.1.2 - but I had forgotten to install the server extension. Crucially though, I had to restart jupyterlab to get them to show up.

**Comment by ggosiang** [comment]
I have the same problem,
JL: 1.1.3
jupy-git: 0.8.1

Not sure which versions are workable?

**Comment by telamonian** [comment]
Closing as resolved. The parts of the code that were giving @mpilosov and @bernhard-42 trouble have been completely refactored since this issue was opened.

If anybody runs into similar problems, please open a new/fresh issue.

**Comment by VolokzhaninVadim** [comment]
I deleted extension and repeated all steps by install. It is solved all my problems.

**Comment by Ark-kun** [comment]
> Closing as resolved. The parts of the code that were giving @mpilosov and @bernhard-42 trouble have been completely refactored since this issue was opened.

The version number that resolves the issue does not seem to be mentioned.

**Comment by fcollonval** [comment]
The refractoring happened in #438 included since 0.9.0
